File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    answer = 0
    info = {'U' : (-1, 0), 'D' : (1, 0), 'R' : (0, 1), 'L' : (0, -1)}
    r, c = 0, 0
    visited = set()
    for d in dirs:
        if -5 <= r + info[d][0] <= 5 and -5 <= c + info[d][1] <= 5:
            dr = r + info[d][0]
            dc = c + info[d][1]
            
            if (r, c, dr, dc) not in visited and (dr, dc, r, c) not in visited:
                print(r, c, dr, dc)
                answer += 1
                visited.add((r, c, dr, dc))
            
            r = dr
            c = dc
            
        else:
            continue
            
    return answer

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.Set;
import java.util.HashSet;

class Solution {
    public int solution(String dirs) {
        Set<String> visited = new HashSet<>();
        
        int curX = 0;
        int curY = 0;
        
        for (int i = 0; i < dirs.length(); i++) {
            int nextX = curX;
            int nextY = curY;
            
            String moving = "";
            
            if (dirs.charAt(i) == 'U') {
                nextY++;
                moving += curX + "" + curY + "->" + nextX + nextY;
            } else if (dirs.charAt(i) == 'D') {
                nextY--;
                moving += nextX + "" + nextY + "->" + curX + curY;
            } else if (dirs.charAt(i) == 'R') {
                nextX++;
                moving += curX + "" + curY + "->" + nextX + nextY;
            } else {
                nextX--;
                moving += nextX + "" + nextY + "->" + curX + curY;
            }
            
            if (nextX < -5 || nextX > 5 || nextY < -5 || nextY > 5) {
                continue;
            }
            visited.add(moving);
            curX = nextX;
            curY = nextY;
        }
        return visited.size();
    }
}